src/main/java/org/apache/commons/text/numbers/ParsedDecimal.java

```diff
@@ -0,0 +1,715 @@
+        int removedDigits = digitCount - count;
+        int i;
+        for (i = count - 1; i >= 0; --i) {
+            final int d = digits[i] + 1;
+
+            if (d < DECIMAL_RADIX) {
+                // value did not carry over; done adding
+                digits[i] = d;
+                break;
+            } else {
+                // value carried over; the current position is 0
+                // which we will ignore by shortening the digit count
+                ++removedDigits;
+            }
+        }
+
+        if (i < 0) {
+            // all values carried over
+            setSingleDigitValue(1, exponent + removedDigits);
+        } else {
+            // values were updated in-place; just need to update the length
+            truncate(digitCount - removedDigits);
+        }
+    }
+
+    /** Return true if this value is equal to zero. The sign field is ignored,
+     * meaning that this method will return true for both {@code +0} and {@code -0}.
+     * @return true if the value is equal to zero
+     */
+    boolean isZero() {
+        return digits[0] == 0;
+    }
+
+    /** Set the value of this instance to a single digit with the given exponent.
+     * The sign of the value is retained.
+     * @param digit digit value
+     * @param newExponent new exponent value
+     */
+    private void setSingleDigitValue(final int digit, final int newExponent) {
+        digits[0] = digit;
+        digitCount = 1;
+        exponent = newExponent;
+    }
+
+    /** Truncate the value to the given number of digits.
+     * @param count number of digits; must be greater than zero and less than
+     *      the current number of digits
+     */
+    private void truncate(final int count) {
+        int nonZeroCount = count;
+        for (int i = count - 1;
+                i >= 0 && digits[i] == 0;
+                --i) {
+            --nonZeroCount;
+        }
+        exponent += digitCount - nonZeroCount;
+        digitCount = nonZeroCount;
+    }
+
+    /** Construct a new instance from the given double value.
+     * @param d double value
+     * @return a new instance containing the parsed components of the given double value
+     * @throws IllegalArgumentException if {@code d} is {@code NaN} or infinite
+     */
+    public static ParsedDecimal from(final double d) {
+        if (!Double.isFinite(d)) {
+            throw new IllegalArgumentException("Double is not finite");
+        }
+
+        // Get the canonical string representation of the double value and parse
+        // it to extract the components of the decimal value. From the documentation
+        // of Double.toString() and the fact that d is finite, we are guaranteed the
+        // following:
+        // - the string will not be empty
+        // - it will contain exactly one decimal point character
+        // - all digit characters are in the ASCII range
+        final char[] strChars = Double.toString(d).toCharArray();
```

Comment:
@darkma773r Did you look at using `String.charAt(int)` in-place of getting a copy of the `char[]`? It would save a trivial amount of memory allocation at the cost of method calls to get each char.

I believe the largest `char[]` required for a double is length 24, e.g.
```java
Double.toString(-Math.nextDown(Math.nextDown(Double.MIN_NORMAL)))
-2.2250738585072004E-308
```
So memory allocation is low. But it if it makes no speed difference then the `charAt` version would have less garbage collection.

Reply:
I switched to using raw char arrays since I profiled the benchmarks and saw a non-trivial amount of time being spent in 'String.charAt()`

Reply:
OK. I did wonder about it as my JDK source code has an index comparison check:
```java
    public char charAt(int index) {
        if ((index < 0) || (index >= value.length)) {
            throw new StringIndexOutOfBoundsException(index);
        }
        return value[index];
    }
```
It would appear that the method call here is significant to performance.